Editors/PathEditor.cpp

```diff
@@ -1,10 +1,238 @@
+    pathModel->setData(Path::kBackgroundRoomNameFieldNumber, 0, "");
+  });
+
+  connect(roomButton->menu, &QMenu::triggered, this, [=](QAction* action) {
+    roomLineEdit->setText(action->text());
+    ui->roomView->SetResourceModel(MainWindow::resourceMap->GetResourceByName(TypeCase::kRoom, action->text())
+                                       ->GetSubModel(TreeNode::kRoomFieldNumber));
+    ui->pathPreviewBackground->SetZoom(1);
+    pathModel->setData(Path::kBackgroundRoomNameFieldNumber, 0, action->text());
+  });
+
+  ui->mainToolBar->addSeparator();
+
+  QToolButton* zoom(new QToolButton(this));
+  zoom->setIcon(ArtManager::GetIcon(":/actions/zoom.png"));
+  ui->mainToolBar->addWidget(zoom);
+  connect(zoom, &QToolButton::pressed, ui->pathPreviewBackground, [=]() { ui->pathPreviewBackground->SetZoom(1); });
+
+  QToolButton* zoomIn(new QToolButton(this));
+  zoomIn->setIcon(ArtManager::GetIcon(":/actions/zoom-in.png"));
+  ui->mainToolBar->addWidget(zoomIn);
+  connect(zoomIn, &QToolButton::pressed, ui->pathPreviewBackground,
+          [=]() { ui->pathPreviewBackground->SetZoom(ui->pathPreviewBackground->GetZoom() * 2); });
+
+  QToolButton* zoomOut(new QToolButton(this));
+  zoomOut->setIcon(ArtManager::GetIcon(":/actions/zoom-out.png"));
+  ui->mainToolBar->addWidget(zoomOut);
+  connect(zoomOut, &QToolButton::pressed, ui->pathPreviewBackground,
+          [=]() { ui->pathPreviewBackground->SetZoom(ui->pathPreviewBackground->GetZoom() / 2); });
+
+  connect(ui->actionShowGrid, &QAction::triggered, [=]() {
+    ui->roomView->GetGrid().show = !ui->roomView->GetGrid().show;
+    ui->pathPreviewBackground->update();
+  });
+
+  connect(ui->actionSnaptoGrid, &QAction::triggered, [this]() { this->snapToGrid = !this->snapToGrid; });
+
+  cursorPositionLabel = new QLabel();
+
+  ui->statusBar->addWidget(cursorPositionLabel);
+
+  ui->pointsTableView->installEventFilter(this);
+
+  connect(xSnap, QOverload<int>::of(&QSpinBox::valueChanged),
+          [this](int val) { ui->roomView->GetGrid().horSpacing = val; });
+
+  connect(ySnap, QOverload<int>::of(&QSpinBox::valueChanged),
+          [this](int val) { ui->roomView->GetGrid().vertSpacing = val; });
+
+  connect(this, &BaseEditor::FocusGained, [=]() { ui->pathPreviewBackground->parentHasFocus = true; });
+  connect(this, &BaseEditor::FocusLost, [=]() { ui->pathPreviewBackground->parentHasFocus = false; });
+
+  nodeMapper->addMapping(ui->nameEdit, TreeNode::kNameFieldNumber);
+
+  resMapper->addMapping(ui->smoothCheckBox, Path::kSmoothFieldNumber);
+  resMapper->addMapping(ui->closedCheckBox, Path::kClosedFieldNumber);
+  resMapper->addMapping(ui->precisionSpinBox, Path::kPrecisionFieldNumber);
+  resMapper->addMapping(xSnap, Path::kHsnapFieldNumber);
+  resMapper->addMapping(ySnap, Path::kVsnapFieldNumber);
+
+  RebindSubModels();
 }
 
 PathEditor::~PathEditor() { delete ui; }
+
+void PathEditor::RebindSubModels() {
+  pathModel = _model->GetSubModel(TreeNode::kPathFieldNumber);

```

Comment:
What prevents this from going stale?

Reply:
The base model (or treenode) is only explicitly deleted by the user. If _model is null the editor should be closed

Reply:
I see. So these are (or at least _model is) a weak pointer, and the editor will simply close when this would otherwise go stale. Sounds fine, although might lead to annoying behavior around, eg, undoing and redoing add or delete operations.